lib/mongoid/indexable/specification.rb

```diff
@@ -63,39 +63,73 @@ def name
 
       private
 
-      # Normalize the spec, in case aliased fields are provided.
+      # Normalize the spec in place, in case aliased fields are provided.
       #
       # @api private
       #
       # @example Normalize the spec.
-      #   specification.normalize_key(name: 1)
+      #   specification.normalize_aliases!(name: 1)
       #
-      # @param [ Hash ] key The index key(s).
+      # @param [ Hash ] spec The index specification.
       #
       # @return [ Hash ] The normalized specification.
-      def normalize_key(key)
-        normalized = {}
-        key.each_pair do |name, direction|
-          normalized[klass.database_field_name(name).to_sym] = direction
+      def normalize_aliases!(spec)
+        return unless spec.is_a?(Hash)
+
+        spec.transform_keys! do |name|
+          klass.database_field_name(name).to_sym
         end
-        normalized
       end
 
-      # Normalize the index options, if any are provided.
+      # Normalize the index options in place.
       #
       # @api private
       #
       # @example Normalize the index options.
-      #   specification.normalize_options(unique: true)
+      #   specification.normalize_options!(unique: true)
       #
-      # @param [ Hash ] opts The index options.
+      # @param [ Hash ] options The index options.
       #
       # @return [ Hash ] The normalized options.
-      def normalize_options(opts)
-        options = {}
-        opts.each_pair do |option, value|
-          options[MAPPINGS[option] || option] = value
+      def normalize_options!(options)
```

Comment:
The reason to do modifications in place is that `recursive_normalize_conditionals!` becomes a lot cleaner; we no longer have to worry about returning values in the recursion tree. Given this design decision, I thought it would be more consistent to use in-place modification everywhere in this class and do the necessary `dup` call at the beginning. Ultimately it makes for more terse code.

Reply:
Personally, I prefer things not to be done in place if it can be avoided, (side note: I think if you simplify the recursion as I have pointed out below, I have a feeling this can be done with a simple `reduce`), however I'm not gonna hold up this PR just because of that.